### PR TITLE
feat: e2e tests

### PR DESCRIPTION
**Pull request by jaredallard**
## What this PR does / why we need it

This PR adds an E2E testing framework for the devenv.

**JIRA ID**: XX-XX

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -14,8 +14,6 @@ jobs:
           machine: true
       - shared/with_go_cache
       - shared/save_go_cache
-      - shared/setup_devenv:
-          provision_args: "--base"
       - run:
```
Comment (jaredallard) [inline .circleci/config.yml:17]:
```suggestion
      - shared/setup_devenv:
          provision: false
      - run:
```

**Comment by jaredallard** [comment]
Will fix tests in the AM. This depends on https://github.com/getoutreach/devbase/pull/140

File: cmd/devenv/devenv.go [hunk 2]
```diff
@@ -82,7 +82,7 @@ func main() {
 				return errors.Wrap(err, "failed to get user home dir")
 			}
 
-			err = os.MkdirAll(filepath.Join(homeDir, ".local", "dev-environment"), 0755)
+			err = os.MkdirAll(filepath.Join(homeDir, ".local", "dev-environment"), 0o755)
```
Comment (ChrisPRobinson) [inline cmd/devenv/devenv.go:85]: nit: typo?

File: cmd/devenv/snapshot/snapshot.go [hunk 3]
```diff
@@ -136,277 +154,341 @@ func NewCmdSnapshot(log logrus.FieldLogger) *cli.Command { //nolint:funlen
...
+		PartitionID:       "aws",
+		URL:               a.endpoint,
+		HostnameImmutable: true,
+		SigningRegion:     a.region,
+	}, nil
+}
 
-	if snapshotName == "" {
-		return fmt.Errorf("missing snapshot name")
+func (o *Options) Generate(ctx context.Context, s *box.SnapshotGenerateConfig,
```
Comment (ChrisPRobinson) [inline cmd/devenv/snapshot/snapshot.go:173]: Nit: The naming here throws me a bit here. What is 'Options' and why does it `Generate`. Reading the code it appears  generate snapshots, push them and set lockfiles, which makes sense given the command this is part of. I don't have a specific solution.

File: cmd/devenv/snapshot/snapshot.go [hunk 4]
```diff
@@ -136,277 +154,341 @@ func NewCmdSnapshot(log logrus.FieldLogger) *cli.Command { //nolint:funlen
...
+	if endpoint := o.b.DeveloperEnvironmentConfig.SnapshotConfig.Endpoint; endpoint != "" {
+		cfg.EndpointResolver = &awsEndpointResolver{ //nolint:staticcheck // Why: using new one doesn't work?
+			endpoint: endpoint,
+			region:   cfg.Region,
+		}
+	}
 
-	return err
-}
+	o.log.WithField("region", cfg.Region).Info("s3 config")
```
Comment (ChrisPRobinson) [inline cmd/devenv/snapshot/snapshot.go:200]: consider adding the endpoint field as well?

File: pkg/kube/logs.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// JobSucceeded returns an error if the given job hasn't succeeded yet.
+// When an error is returned, this is considered unrecoverable.
+func JobSucceeded(ctx context.Context, k kubernetes.Interface, name, namespace string) (bool, error) {
```
Comment (ChrisPRobinson) [inline pkg/kube/logs.go:25]: Nice function!

File: cmd/devenv/snapshot/snapshot.go [hunk 6]
```diff
@@ -136,277 +154,341 @@ func NewCmdSnapshot(log logrus.FieldLogger) *cli.Command { //nolint:funlen
...
 	}
+	destroyOpts.Run(ctx) //nolint:errcheck
 
-	if liveRestore {
-		o.log.Warn("THIS WILL DELETE ALL EXISTING DATA IN YOUR CLUSTER FROM BEFORE THE SNAPSHOT. PROCEED?")
-		proceed, err := cmdutil.GetYesOrNoInput(ctx)
-		if err != nil {
-			return err
-		}
+	// using exec because of an import cycle, need to fix
```
Comment (ChrisPRobinson) [inline cmd/devenv/snapshot/snapshot.go:399]: What is the import cycle itself?

File: cmd/devenv/snapshot/snapshot.go (hunk 3 above)
Comment (jaredallard) [reply cmd/devenv/snapshot/snapshot.go:173]: Yeah, the naming is odd. It's kept as an artifact. Really it's options to the command itself in an attempt to be more testable. cli.Context -> creates Options, runs that. It needs to be fixed down the line for sure.

File: cmd/devenv/snapshot/snapshot.go (hunk 6 above)
Comment (jaredallard) [reply cmd/devenv/snapshot/snapshot.go:399]: Ah, this can be removed now. Old comment. I fixed the import cycle.

Comment (jaredallard) [inline cmd/devenv/snapshot/snapshot.go:399]:
```suggestion
```

File: pkg/kube/logs.go (hunk 5 above)
Comment (jaredallard) [inline pkg/kube/logs.go:25]:
```suggestion
// JobSucceeded returns a bool if the given job has/hasn't succeeded.
// If an error is returned, this is considered unrecoverable.
func JobSucceeded(ctx context.Context, k kubernetes.Interface, name, namespace string) (bool, error) {
```

File: internal/e2e/e2e_test.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	if err := devenv.DestroyDevenv(ctx, provisionArgs.Box); err != nil {
+		t.Error(errors.Wrap(err, "failed to destroy devenv after snapshot generation"))
+		return
+	}
+
+	provisionArgs.SnapshotTarget = "default"
+
+	t.Log("Generating devenv using generated snapshot")
+	devenv.ProvisionDevenv(t, context.Background(), &provisionArgs)()
+}
```
[the remark this answers is not in the document]
Comment (jaredallard) [reply internal/e2e/e2e_test.go:108]: Nope, this is it. It's also technically not needed. This is nuked between snapshots. Was just a best effort thing!

File: pkg/snapshot/from_snapshot.go [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+	veleroclient "github.com/vmware-tanzu/velero/pkg/generated/clientset/versioned"
+	velerov1 "github.com/vmware-tanzu/velero/pkg/generated/informers/externalversions/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/util/boolptr"
+)
+
+const (
+	SnapshotNamespace = "velero"
+)
+
+type Options struct {
```
[the remark this answers is not in the document]
Comment (jaredallard) [reply pkg/snapshot/from_snapshot.go:35]: Yes! I can rename this, I should've when I moved it out of `cmd/snapshot`

File: pkg/snapshot/from_snapshot.go [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+		opts.vc, err = veleroclient.NewForConfig(conf)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create snapshot client")
+		}
+	}
+
+	return opts, nil
+}
+
+func (o *Options) GetSnapshot(ctx context.Context, snapshotName string) (*velerov1api.Backup, error) {
```
[the remark this answers is not in the document]
Comment (ChrisPRobinson) [reply pkg/snapshot/from_snapshot.go:69]: or something like `backup.Retrieve`

File: pkg/snapshot/from_snapshot.go [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+					restore, ok := obj.(*velerov1api.Restore)
+					if !ok {
+						return
+					}
+					updates <- restore
+				},
+			},
+		},
+	)
+	go restoreInformer.Run(ctx.Done())
```
Comment (ChrisPRobinson) [inline pkg/snapshot/from_snapshot.go:196]: Does this happen quickly? How do we know when this has completed? So does this mean restore is async?

**Review by ChrisPRobinson — APPROVED** [review]
LGTM

**Pull request merged**